### PR TITLE
Allow arguments other than Time in FFT

### PR DESCRIPTION
**Pull request by eggrobin**
This will be used in Clenshaw-Curtis, which in turn will be used to compute the inner products and norms in the Modified Gram-Schmidt introduced by #2750, continuing the inexorable progress towards #2400.

### CODE REVIEW

File: numerics/fast_fourier_transform.hpp [hunk 1]
```diff
@@ -21,46 +21,59 @@ using base::FloorLog2;
...
+
+// Given (u₀, ..., uₙ), this class computes the discrete Fourier transform
+//   Uₛ = ∑ᵣ uᵣ exp(-2πirs/n),
+// corresponding to
+//   Fourier[{...}, FourierParameters -> {1, -1}]
+// in Mathematica notation (the "signal processing" Fourier transform).
+template<typename Value, typename Argument, std::size_t size_>
 class FastFourierTransform {
  public:
+  using AngularFrequency = Derivative<Angle, Argument>;
```
Comment (pleroy) [inline numerics/fast_fourier_transform.hpp:36]: Document that this is only a *bona fide* angular frequency if `Argument` is time-like.

File: numerics/fast_fourier_transform.hpp [hunk 2]
```diff
@@ -21,46 +21,59 @@ using base::FloorLog2;
...
 
+  // Returns the coefficient Uₛ.
+  Complexification<Value> const& operator[](int s) const;
+
+  // Returns the frequency corresponding to Uₛ.
+  AngularFrequency frequency(int s) const;
+
  private:
-  Time const Δt_;
+  Difference<Argument> const Δt_;
```
Comment (pleroy) [inline numerics/fast_fourier_transform.hpp:76]: Document that this is only a *bona fide* time if `Argument` is time-like.

File: numerics/fast_fourier_transform_body.hpp [hunk 3]
```diff
@@ -143,15 +143,17 @@ FastFourierTransform<Value, size_>::FastFourierTransform(
...
+    Difference<Argument> const& Δt)
     : FastFourierTransform(container.cbegin(), container.cend(), Δt) {}
 
-template<typename Value, std::size_t size_>
-std::map<AngularFrequency, typename Hilbert<Value>::InnerProductType>
-FastFourierTransform<Value, size_>::PowerSpectrum() const {
+template<typename Value, typename Argument, std::size_t size_>
+std::map<
+    typename FastFourierTransform<Value, Argument, size_>::AngularFrequency,
+    typename Hilbert<Value>::InnerProductType>
```
Comment (pleroy) [inline numerics/fast_fourier_transform_body.hpp:155]: I would not object to a trailing return type here.

File: numerics/fast_fourier_transform_body.hpp [hunk 4]
```diff
@@ -162,8 +164,10 @@ FastFourierTransform<Value, size_>::PowerSpectrum() const {
   return spectrum;
 }
 
-template<typename Value, std::size_t size_>
-Interval<AngularFrequency> FastFourierTransform<Value, size_>::Mode() const {
+template<typename Value, typename Argument, std::size_t size_>
+Interval<
+    typename FastFourierTransform<Value, Argument, size_>::AngularFrequency>
```
Comment (pleroy) [inline numerics/fast_fourier_transform_body.hpp:169]: Same.

File: numerics/fast_fourier_transform_body.hpp [hunk 5]
```diff
@@ -187,6 +191,19 @@ Interval<AngularFrequency> FastFourierTransform<Value, size_>::Mode() const {
...
+template<typename Value, typename Argument, std::size_t size_>
+Complexification<Value> const&
+    FastFourierTransform<Value, Argument, size_>::operator[](
+        int const s) const {
+  return transform_[s];
+}
+
+template<typename Value, typename Argument, std::size_t size_>
+typename FastFourierTransform<Value, Argument, size_>::AngularFrequency
+FastFourierTransform<Value, Argument, size_>::frequency(int const s) const {
```
Comment (pleroy) [inline numerics/fast_fourier_transform_body.hpp:203]: `DCHECK` that `s` is in the correct range.

File: numerics/fast_fourier_transform.hpp [hunk 6]
```diff
@@ -21,46 +21,59 @@ using base::FloorLog2;
...
   PowerSpectrum() const;
 
   // Returns the interval that contains the largest peak of power.
   Interval<AngularFrequency> Mode() const;
 
+  // Returns the coefficient Uₛ.
+  Complexification<Value> const& operator[](int s) const;
+
+  // Returns the frequency corresponding to Uₛ.
+  AngularFrequency frequency(int s) const;
```
Comment (pleroy) [inline numerics/fast_fourier_transform.hpp:73]: In this function and the next, document the acceptable values of `s`.

File: numerics/fast_fourier_transform_test.cpp [hunk 7]
```diff
@@ -190,6 +198,23 @@ TEST_F(FastFourierTransformTest, Vector) {
               AlmostEquals(4 * π / FFT::size * Radian / Second, 24));
 }
 
+TEST_F(FastFourierTransformTest, Inverse) {
+  constexpr int n = 4;
+  constexpr Time Δt = 1 * Second;
+  std::array<Voltage, n> u{{1 * Volt, 0 * Volt, 1 * Volt, 0 * Volt}};
+  FastFourierTransform<Voltage, Instant, n> const U(u, Δt);
+
+  FastFourierTransform<Voltage, AngularFrequency, n> const nu(
```
Comment (pleroy) [inline numerics/fast_fourier_transform_test.cpp:207]: Use Greek nu?

File: numerics/fast_fourier_transform_test.cpp [hunk 8]
```diff
@@ -190,6 +198,23 @@ TEST_F(FastFourierTransformTest, Vector) {
...
+      {U[0].real_part(), U[1].real_part(), U[2].real_part(), U[3].real_part()},
+      U.frequency(1) - U.frequency(0));
+  EXPECT_THAT((std::array{nu[0].real_part() / n,
+                          nu[1].real_part() / n,
+                          nu[2].real_part() / n,
+                          nu[3].real_part() / n}),
+              ElementsAreArray(u));
+  EXPECT_THAT(nu.frequency(1) - nu.frequency(0), Eq(Δt));
```
Comment (pleroy) [inline numerics/fast_fourier_transform_test.cpp:215]: Use `AlmostEquals` instead of `Eq`.

**Comment by pleroy** [comment]
retest this please

File: numerics/fast_fourier_transform_test.cpp (hunk 7 above)
Comment (eggrobin) [reply numerics/fast_fourier_transform_test.cpp:207]: No, this is n times u. Renamed u to v for clarity.

File: numerics/fast_fourier_transform.hpp (hunk 2 above)
Comment (eggrobin) [reply numerics/fast_fourier_transform.hpp:76]: This is what the type says (and the word time does not appear anyway).

**Pull request merged**